src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,505 @@
+            {
+                return null;
+            }
+        }
+
+        private JsonTree ParseTree(bool strict)
+        {
+            var arraySequence = this.ParseSequence();
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(_currentToken.Kind == JsonKind.EndOfFile);
+
+            var root = new JsonCompilationUnit(arraySequence, _currentToken);
+
+            // We only report a single diagnostic when parsing out json.  This helps prevent lots of
+            // cascading errors from being reported.  First, we see if there are any diagnostics
+            // directly in tokens in the tree.  If not, we then check for any incorrect tree
+            // structure (that would be incorrect for both json.net or strict-mode).  If we don't
+            // run into any problems, we'll then perform specific json.net or strict-mode checks.
+            var diagnostic = GetFirstDiagnostic(root) ?? CheckTopLevel(_lexer.Text, root);
+
+            if (diagnostic == null)
+            {
+                // We didn't have any diagnostics in the tree so far.  Do the json.net/strict checks
+                // depending on how we were invoked.
+                diagnostic = strict
+                    ? StrictSyntaxChecker.CheckSyntax(root)
+                    : JsonNetSyntaxChecker.CheckSyntax(root);
+            }
+
+            var diagnostics = diagnostic == null
+                ? ImmutableArray<EmbeddedDiagnostic>.Empty
+                : ImmutableArray.Create(diagnostic.Value);
+
+            return new JsonTree(
+                _lexer.Text, root, diagnostics);
+        }
+
+        /// <summary>
+        /// Checks for errors in json for both json.net and strict mode.
+        /// </summary>
+        private static EmbeddedDiagnostic? CheckTopLevel(
+            VirtualCharSequence text, JsonCompilationUnit compilationUnit)
+        {
+            var arraySequence = compilationUnit.Sequence;
+            if (arraySequence.IsEmpty)
+            {
+                // json is not allowed to be just whitespace.
+                if (text.Length > 0 &&
+                    compilationUnit.EndOfFileToken.LeadingTrivia.All(
+                        t => t.Kind == JsonKind.WhitespaceTrivia || t.Kind == JsonKind.EndOfLineTrivia))
+                {
+                    return new EmbeddedDiagnostic(FeaturesResources.Syntax_error, GetSpan(text));
+                }
+            }
+            else if (arraySequence.Length >= 2)
+            {
+                // the top level can't have more than one actual value.
+                var firstToken = GetFirstToken(arraySequence[1]);
+                return new EmbeddedDiagnostic(
+                    string.Format(FeaturesResources._0_unexpected, firstToken.VirtualChars[0]),
+                    firstToken.GetSpan());
+            }
+
+            foreach (var child in compilationUnit.Sequence)
+            {
```

Comment:
Why the foreach? At this point, it's a single element, right?

Reply:
touche :)